src/libsystemd/sd-device/sd-device.c

```diff
@@ -2326,123 +2326,207 @@ void device_clear_sysattr_cache(sd_device *device) {
-         * This replaces the value if it already exists. */
+        /* This takes the reference of the input arguments on success. The input value may be NULL.
+         * This replaces an already existing entry. */
 
-        /* First, remove the old cache entry. So, we do not need to clear cache on error. */
-        old_value = hashmap_remove2(device->sysattr_values, key, (void **) &new_key);
-        if (!new_key) {
-                new_key = strdup(key);
-                if (!new_key)
-                        return -ENOMEM;
-        }
+        /* Remove the old cache entry. So, we do not need to clear cache on error. */
+        sysattr_cache_entry_free(hashmap_remove(device->sysattr_values, key));
+
+        /* We use ENOANO as a recognizable error code when we have not read the attribute. */
+        if (error == ENOANO)
+                error = ESTALE;
+
+        _cleanup_free_ SysAttrCacheEntry *entry = new(SysAttrCacheEntry, 1);
+        if (!entry)
+                return -ENOMEM;
 
-        r = hashmap_ensure_put(&device->sysattr_values, &path_hash_ops_free_free, new_key, value);
+        *entry = (SysAttrCacheEntry) {
+                .key = key,
+                .value = value,
+                .error = error,
+        };
+
+        r = hashmap_ensure_put(&device->sysattr_values, &sysattr_cache_hash_ops, entry->key, entry);
         if (r < 0)
                 return r;
 
-        TAKE_PTR(new_key);
-
+        TAKE_PTR(entry);
         return 0;
 }
 
 int device_get_cached_sysattr_value(sd_device *device, const char *key, const char **ret_value) {
-        const char *k = NULL, *value;
+        SysAttrCacheEntry *entry;
 
         assert(device);
         assert(key);
 
-        value = hashmap_get2(device->sysattr_values, key, (void **) &k);
-        if (!k)
-                return -ESTALE; /* We have not read the attribute. */
-        if (!value)
-                return -ENOENT; /* We have looked up the attribute before and it did not exist. */
+        entry = hashmap_get(device->sysattr_values, key);
+        if (!entry)
+                return -ENOANO; /* We have not read the attribute. */
+        if (!entry->value) {
+                /* We have looked up the attribute before and failed. Return the cached error code. */
+                assert(entry->error > 0);
+                return -entry->error;
+        }
         if (ret_value)
-                *ret_value = value;
+                *ret_value = entry->value;
+        return 0;
+}
+
+int device_chase(sd_device *device, const char *path, ChaseFlags flags, char **ret_resolved, int *ret_fd) {
+        int r;
+
+        assert(device);
+        assert(path);
+
+        const char *syspath;
+        r = sd_device_get_syspath(device, &syspath);
+        if (r < 0)
+                return r;
+
+        /* Here, CHASE_PREFIX_ROOT is borrowed. if it is set, the specified path will be prefixed with the syspath. */
```

Comment:
maybe mention that doing it this way means absolute symlinks remain absolute

Reply:
Sorry, I cannot follow ???

Reply:
Or, I may misunderstand your point, but maybe better to unconditionally prefix with syspath unless the specified path does not start with `/sys/` ? Then, we can drop the 'spurious' borrowing.

Reply:
what i wanted to say, is that chase() with the root path specified as 2nd argument would mean (if CHASE_PREFIX_ROOT is specified) that we'd concat the root and path automatically, but it would *also* have the effect that absolute symlinks would be resolved relative to that root, instead of the real host root. hence it makes sense to join the paths beforehand so that we get the joined paths but chase() will not use it for resolving absolute symliks.

this is mostly theoretical, because /sys/ hopefully doesn't have any absolute symlinks (but then again /proc/ has some to /sys/ for the devicetree stuff iirc). but i still think this complex should be briefly mentioned in a comment.